Submenu: Add revert button.

getdave opened the pull request:
<img width="657" alt="Screen Shot 2022-02-01 at 10 47 41" src="https://user-images.githubusercontent.com/444434/151955187-810d1780-5e1c-4cc9-ab93-45a46d8ea879.png">



## Description
In https://github.com/WordPress/gutenberg/issues/36977#issuecomment-1020225327 we learnt that some users want to be able to revert the submenu block to a standard link. The expectation was/is that deleting the submenu would restore the link but of course as the block is in fact a submenu block this just removes the entire block.

This PR seeks to implement one possible solution which is to show an explicit "Revert to Link" button in the block toolbar. This effectively just does the same thing as the block's built-in "Transform" option, but it makes it more explicit for this specific use case.

I'm no designer so this is just one idea for solving this case and we will also no doubt need a better icon choice. 

Curious to hear from those more experienced with this block as well.

Also would love feedback from @courtneyr-dev who originally noticed the quirk and requested the change.

## How has this been tested?

1. Add Nav block.
2. Create items.
3. Add a submenu to one of the items and add some sub items.
4. On the parent submenu item, click the toolbar button to revert to a standard link.
5. The link details should all be retained except from any child blocks which will be disguarded.

⚠️ As noted above, any children of the submenu block being revert will be lost. Is that the behaviour we want? If not then what should happen?

## Screenshots <!-- if applicable -->


https://user-images.githubusercontent.com/444434/151781990-ab0a4c1b-2a37-4b36-b977-9210796746b8.mp4



## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

getdave committed:
Add revert button

talldan commented on packages/block-library/src/navigation-submenu/edit.js:
```diff
@@ -542,6 +550,14 @@ export default function NavigationSubmenuEdit( {
+					<ToolbarButton
+						name="revert"
+						icon={ keyboardReturn }
+						title={ __( 'Revert to Link' ) }
```
How about just 'Convert to Link'?

I'm thinking about a case where a user adds a submenu directly from the global inserter, in which case it would be a reversion.

getdave replied:
✅ Sounds good! Done.

getdave committed:
Adjust button wording

tellthemachines reviewed (commented):
I'm hesitant to enable this without showing a warning when there are items in the submenu. As it is, it's too easy to delete a whole submenu by clicking on a button with an icon that looks like a simple "undo" action.

Even if we show the text label, it might be confusing, particularly if "Open on click" is enabled (users might think they're just converting the top-level item back into a link instead of getting rid of the whole submenu).

getdave commented:
I recognise the concerns expressed here. That said, we have a clear use case where users are being confused by the continued presence of a submenu block even when there are no submenu items. I'd really like to find a way to smooth out that path.

I appreciate your help.

> I'm hesitant to enable this without showing a warning when there are items in the submenu. 

That's a fair point. Would adding a confirmation dialogue if there are sub items resolve this concern for you? If not is there anything else you can think of that would?

courtneyr-dev commented:
>  Users are being confused by the continued presence of a submenu block even when there are no submenu items.

I concur. If users leave the parent menu as having a submenu when none are present, website visitors would also be confused seeing the drop-down with no submenu items present. 

I am not the best yet with code, but the idea would be, if there is 1+, then show dropdown, else show standard link.

TimothyBJacobs commented:
Is it feasible to make deleting a block function as the users expect? That has the advantages of users not needing to learn another pattern for destructive actions. 


I think it'd also be generally useful as well to allow for blocks to provide an `onDelete` handler of some kind.

talldan commented:
> Is it feasible to make deleting a block function as the users expect? That has the advantages of users not needing to learn another pattern for destructive actions.

@TimothyBJacobs Would you be able to go into more detail about this?

> I'm hesitant to enable this without showing a warning when there are items in the submenu.

It could also be that the button only appears when the submenu is empty so that there's less risk.

I think there's also the possibility for more natural in-canvas options for add/remove submenus. Like a `+` button below a link to create a submenu (it could work like the in-between inserter). Similarly an `x` button in the same place to remove an empty submenu. This is just a random idea though, and it could do with some design exploration.

getdave commented:
> It could also be that the button only appears when the submenu is empty so that there's less risk.

I think that's a good simple solution @talldan. Thank you for the clear insight.

> Is it feasible to make deleting a block function as the users expect? That has the advantages of users not needing to learn another pattern for destructive actions.

I don't think it's possible at this time and, whilst I understand the motivation, I'm also not convinced it's something we should be doing. The delete action always performs the same behaviour across all blocks - removing the block. If we start making exceptions we could end up with a fragmented and confusing user experience.

I would still vote for a clear button dedicated to this edge case (as per this PR). I'm going to refine the PR today to include some of the suggestions included here whilst avoiding any "magic" behaviour (i.e. things happening automatically based on assumptions about user intent).

I hope you don't mind me adding a few of you as reviewers on the PR? I'll ping you when it's ready. 

Thanks again for all the feedback and input here. It's really productive and useful 🙇‍♂️

getdave committed:
Only show convert button when no children present

getdave committed:
Move to separate menu group

getdave commented:
Ok @talldan @courtneyr-dev @TimothyBJacobs I've now changed the code so that the button only appears is there are no items in the submenu. That feels like a healthy compromise.

I've updated the design to use the `undo` icon but I'd love to hear from @javierarce on any suggested alternatives to this.

getdave committed:
USe undo icon

draganescu commented:
Random thought: what if instead of yet-another-button we move to do things preemptively:

- add a warning to the block's inspector, like "you have empty submenus" and an action to "Convert empty submenus to links"
- automatically convert a submenu (a la autocorrect) to a link when the last child of the submenu is removed. This I think is unobtrusive and easily reverted by the user because adding a child converts the link to a submenu again.

On the PR in particular, the icon absolutely has to be changed. I do think that showin that button only when the submenu is empty is a great idea, but my suggestions above are an attempt at a more fluid experience.

getdave commented:
> * add a warning to the block's inspector, like "you have empty submenus" and an action to "Convert empty submenus to 

That's another option I'd not considered. More "hidden" though.

> * automatically convert a submenu (a la autocorrect) to a link when the last child of the submenu is removed. This I think is unobtrusive and easily reverted by the user because adding a child converts the link to a submenu again.

Some [prior discussion on this here](https://github.com/WordPress/gutenberg/issues/36977#issuecomment-1020748782). There was no clear consensus but in general I dislike "magic" in UIs. Generally I prefer clear unambiguous controls that respond to user action. Automatically converting the submenu back to a link might also be a confusing experience, but I'm more than open to alternatives - I might even open a PR for comparison...perhaps that will be best!


> On the PR in particular, the icon absolutely has to be changed. 

I agree as I mentioned in the description 😄 

>  I do think that showin that button only when the submenu is empty is a great idea...

Kudos to @talldan for that one.


> ...but my suggestions above are an attempt at a more fluid experience.

I recognise this. I still have concerns about having things happen "by magic" but I'd like to explore this direction as well.

javierarce commented:
> I've updated the design to use the undo icon but I'd love to hear from @javierarce on any suggested alternatives to this.

As an alternative solution, maybe we could create a new icon using the same idea as the unlink one.

![image](https://user-images.githubusercontent.com/4933/151794274-46c26505-ef62-4599-aa52-976392ccb976.png)

And I would also change the order of the buttons so the on/off action is more clear.

![image](https://user-images.githubusercontent.com/4933/151795170-2fbda628-4a0b-4adb-b054-de30e009f56d.png)

getdave commented:
> > I've updated the design to use the undo icon but I'd love to hear from @javierarce on any suggested alternatives to this.
> 
> As an alternative solution, maybe we could create a new icon using the same idea as the unlink one.
> 
> ![image](https://user-images.githubusercontent.com/4933/151794274-46c26505-ef62-4599-aa52-976392ccb976.png)
> 
> And I would also change the order of the buttons so the on/off action is more clear.
> 
> ![image](https://user-images.githubusercontent.com/4933/151795170-2fbda628-4a0b-4adb-b054-de30e009f56d.png)

Make sense @javierarce - thank you. Can you provide me with the SVG for this new icon?

javierarce commented:
Here you go @getdave, the remove-submenu.svg:

![remove-submenu](https://user-images.githubusercontent.com/4933/151946384-2c9c4794-1716-4b58-81cd-97c0b5d8d413.svg)

I've also added the icon to the [Explored icons column in the Figma library](https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=10519%3A0)

I would appreciate a second opinion/quick check from @jasmussen 🙏

jasmussen commented:
Looks great! 

One thing I like about Material icons is that their strike-out versions only add the negative space shadow on one side of the strike, like so:
<img width="135" alt="Screenshot 2022-02-01 at 10 59 26" src="https://user-images.githubusercontent.com/1204802/151948276-38503307-dfca-4a71-8754-ef40614b9fcf.png">
<img width="111" alt="Screenshot 2022-02-01 at 10 59 32" src="https://user-images.githubusercontent.com/1204802/151948278-ce024e95-2b7d-4e29-bc16-0332efebfc09.png">

I took a quick stab applying the same pattern here:

<img width="764" alt="Screenshot 2022-02-01 at 10 58 57" src="https://user-images.githubusercontent.com/1204802/151948319-0f3391b8-6655-42cb-afa5-870c66197c2d.png">

Depending on which you prefer, here's the SVG:

```
<svg width="24" height="25" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="m13.955 20.748 8-17.5-.91-.416L19.597 6H13.5v1.5h5.411l-1.6 3.5H13.5v1.5h3.126l-1.6 3.5H13.5l.028 1.5h.812l-1.295 2.832.91.416ZM17.675 16l-.686 1.5h4.539L21.5 16h-3.825Zm2.286-5-.686 1.5H21.5V11h-1.54ZM2 12c0 3.58 2.42 5.5 6 5.5h.5V19l3-2.5-3-2.5v2H8c-2.48 0-4.5-1.52-4.5-4S5.52 7.5 8 7.5h3.5V6H8c-3.58 0-6 2.42-6 6Z" /></svg>
```

javierarce commented:
Oh, good point, @jasmussen, thanks! I prefer your version… it's cleaner and easy to read.

getdave committed:
Add new removeSubmenu icon

getdave committed:
Apply icon to revert button

getdave commented:
@jasmussen @javierarce I've added the new icon and applied to the button. I've also moved the icons around to match @javierarce's instructions.

How does it look to you now?

<img width="657" alt="Screen Shot 2022-02-01 at 10 47 41" src="https://user-images.githubusercontent.com/444434/151955152-989a1db6-1673-4fe8-8d06-506dfaf0949f.png">

Estelle544 commented on packages/block-library/src/navigation-submenu/edit.js:
```diff
@@ -38,9 +38,10 @@ import {
 import { store as coreStore } from '@wordpress/core-data';
 import { speak } from '@wordpress/a11y';
+import { createBlock } from '@wordpress/blocks';
 
```
packages/block-library/src/navigation-submenu/edit.js

getdave replied:
Hey thanks for review. Was there something you were looking to highlight here?

javierarce commented:
> How does it look to you now?

It looks good to me!

> I've now changed the code so that the button only appears is there are no items in the submenu. That feels like a healthy compromise.

I found a case where the submenu is not empty (because there are one or more placeholders inside) and the undo button doesn't appear. I don't know if this is complicated to detect, but in this case, it would be great if we could show the undo button.

<img width="418" alt="Screenshot 2022-02-02 at 08 52 39" src="https://user-images.githubusercontent.com/4933/152114465-d87817d8-d778-4957-a2e4-d69920c7a3fb.png">

getdave committed:
Show revert button for single empty link block

getdave commented:
> I found a case where the submenu is not empty (because there are one or more placeholders inside) and the undo button doesn't appear. I don't know if this is complicated to detect, but in this case, it would be great if we could show the undo button.


@javierarce I've updated the code but this has made things a fair bit more verbose. The revert button will now show if:

* There are no items in the submenu.
* There is a single `core/navigation-link` item in the submenu and it has no `url` attribute.

To test this you'll need to try variations of adding empty/non-empty blocks into the submenu. Also worth trying various levels of nesting. It seemed to work to me.

This is probably an edge case but I can see how users might end up in this state.

Also I noticed that you can add a valid link block (i.e. with a URL) but then if you _remove_ the _text_ from the link the placeholder is "Add a link". That's wrong because we already have a "link" because there is a URL. IN such circumstances it should say "Enter link text" or something. Do you agree? This would be one for a separate PR.

adamziel commented:
My ideal option would be what @draganescu mentioned:

> automatically convert a submenu (a la autocorrect) to a link when the last child of the submenu is removed. This I think is unobtrusive and easily reverted by the user because adding a child converts the link to a submenu again.

I understand there is no clear consensus around that though. A revert icon sounds like a good way to proceed here 🚢

adamziel commented on packages/block-library/src/navigation-submenu/edit.js:
```diff
@@ -312,20 +315,39 @@ export default function NavigationSubmenuEdit( {
+				const singleBlock = getBlock( selectedBlockDescendants[ 0 ] );
+
+				_onlyDescendantIsEmptyLink =
+					singleBlock?.name === 'core/navigation-link' &&
```
Nit: this is a very specific condition reasoning about a details of a different block. I fear that a change in either the navigation-link logic or the default placeholder choice would break this in a way that's hard to notice. An E2E test should help. By the way, are there any similar checks in other places?

getdave replied:
Yes I don't love the tight coupling here.

Options:

1. Ask @javierarce how important he feels [this is to UX](https://github.com/WordPress/gutenberg/pull/38203#issuecomment-1027672948) and potentially look to revert.
2. Add an e2e test to cover the scenario.

javierarce replied:
Since we don't allow to revert submenus that contain items, the importance of that UI is probably limited and more of a helpful shortcut… but that said, I think is a nice shortcut because removing links is a bit tricky right now.

getdave replied:
@adamziel I've added e2e test coverage here.

jasmussen commented:
> I understand there is no clear consensus around that though. 

I also tend to want to start simple, but it's not a strong opinion.

javierarce commented:
> Also I noticed that you can add a valid link block (i.e. with a URL) but then if you remove the text from the link the placeholder is "Add a link". That's wrong because we already have a "link" because there is a URL. IN such circumstances it should say "Enter link text" or something. Do you agree? This would be one for a separate PR.

+1. And maybe we should think about changing the link icon too? I wonder if we should use the pencil one instead.

<img width="466" alt="Screenshot 2022-02-07 at 16 05 41" src="https://user-images.githubusercontent.com/4933/152814396-ccce4e8e-6c06-4627-8c60-b30d84b2fddb.png">

draganescu commented:
This tests  very well!

getdave commented:
Next step: add an e2e test to cover this code.

getdave committed:
Add test for reverting empty submenu

getdave committed:
Extract waitForBlock helper function

getdave committed:
Add test for when convert button should not show

getdave committed:
Refactor block toolbar selections to use helper

getdave committed:
Add test to cover single unpopulated link item in submenu

getdave committed:
Increase specificity of block toolbar selector

getdave committed:
Revert extension of populateWithOneItem helper

This is no longer required

getdave committed:
Remove debugging

getdave committed:
Make use of existing e2e test helper

getdave committed:
Correct spelling in comment

getdave commented:
> My ideal option would be what @draganescu mentioned:

@adamziel I responded to that [in this comment](https://github.com/WordPress/gutenberg/pull/38203#issuecomment-1025673023). Hopefully that explains my reasoning around at least trialing this more explicit approach as a first option.

getdave commented:
Assuming the e2e tests pass on CI I'm now seeking a technical ✅ on this PR.

adamziel commented on packages/block-library/src/navigation-submenu/edit.js:
```diff
@@ -529,6 +550,11 @@ export default function NavigationSubmenuEdit( {
 
 	const ParentElement = openSubmenusOnClick ? 'button' : 'a';
 
+	function transformToLink() {
```
wrapping this in `useCallback` could help avoid some re-rendering

getdave replied:
Did you notice a performance problem here? If not I generally [adhere to KCD](https://kentcdodds.com/blog/usememo-and-usecallback#why-is-usecallback-worse).

adamziel commented on packages/block-library/src/navigation-submenu/edit.js:
```diff
@@ -312,20 +315,39 @@ export default function NavigationSubmenuEdit( {
+
+				_onlyDescendantIsEmptyLink =
+					singleBlock?.name === 'core/navigation-link' &&
+					! singleBlock?.attributes?.url;
```
I added a sub-menu item, then removed the label and so it appears "empty." I can't remove it, though – it still has a URL. Would a label be a better attribute to check here?

getdave replied:
🤔 Not sure. Happy to try it. We can always revert it later.

getdave replied:
✅

adamziel commented:
It tests well, thank you @getdave! That being said, I have a UX concern: As a user I am confused why is the "revert" button sometimes not there. Could we always display it but make it inactive with an `alt` text explaining the situation when the submenu isn't "empty"?

getdave commented:
> It tests well, thank you @getdave! That being said, I have a UX concern: As a user I am confused why is the "revert" button sometimes not there. Could we always display it but make it inactive with an `alt` text explaining the situation when the submenu isn't "empty"?

Good question but that's one I'd prefer to defer to an expert such as @javierarce.

> Could we always display it but make it inactive with an alt text explaining the situation when the submenu isn't "empty"?

Also one I'd appreciate @javierarce's opinion on. @adamziel are there any other examples of this pattern in Gutenberg right now? I can't think of one.

adamziel commented:
@getdave I'm not sure! I'm just biased towards _not hiding_ UI elements without an explanation – otherwise I find myself thinking _wait, there was a feature here, hmm... I guess they removed it_.

getdave committed:
Move check for "empty link block" to be based on label

javierarce commented:
> It tests well, thank you @getdave! That being said, I have a UX concern: As a user I am confused why is the "revert" button sometimes not there. Could we always display it but make it inactive with an alt text explaining the situation when the submenu isn't "empty"?

I like @adamziel's idea. It will probably help reduce the frustration.

> Also one I'd appreciate @javierarce's opinion on. @adamziel are there any other examples of this pattern in Gutenberg right now? I can't think of one.

I don't remember examples that are identical… the closest thing could be these disabled buttons in the list toolbar.

![world local_wp-admin_post php_post=1 action=edit](https://user-images.githubusercontent.com/4933/153398145-f63a51b3-89df-4988-a8ba-bc436358193e.png)

getdave commented:
Next step: add disabled button and update e2e tests.

getdave committed:
Disable convert button instead of removing

getdave commented:
<img width="623" alt="Screen Shot 2022-02-10 at 13 19 43" src="https://user-images.githubusercontent.com/444434/153416347-cd328ad3-22c2-4829-a3db-67ef8c3a41ea.png">

@javierarce @adamziel I've updated with the disabled state.

If you're happy with the result (and the code) I'm seeking a ✅

getdave committed:
Fix SVG DOM error by using camelCase prop names

javierarce commented:
> If you're happy with the result (and the code) I'm seeking a ✅

Looks good to me!

draganescu reviewed (approved):
Codewise this LGTM 🚢
